From 37508b45f7591f3acb749afb223aeefcaeb76749 Mon Sep 17 00:00:00 2001
From: Marko Lindqvist <cazfi74@gmail.com>
Date: Wed, 27 Dec 2023 15:11:56 +0200
Subject: [PATCH 17/17] Fix city removal server crashes

When city removal wipes a unit that's not homed to the city,
but on the same tile, recursive city refresh was crashing
due to city vision and advisor data being already deleted.

See RM #81

Signed-off-by: Marko Lindqvist <cazfi74@gmail.com>
---
 server/citytools.c | 17 +++++++++++------
 1 file changed, 11 insertions(+), 6 deletions(-)

diff --git a/server/citytools.c b/server/citytools.c
index f760deb3e0..87917e0b77 100644
--- a/server/citytools.c
+++ b/server/citytools.c
@@ -3416,11 +3416,13 @@ void city_landlocked_sell_coastal_improvements(struct tile *ptile)
 ****************************************************************************/
 void city_refresh_vision(struct city *pcity)
 {
-  v_radius_t vision_radius_sq
-    = V_RADIUS(get_city_bonus(pcity, EFT_CITY_VISION_RADIUS_SQ), 2, 2);
+  if (pcity->server.vision != nullptr) {
+    v_radius_t vision_radius_sq
+      = V_RADIUS(get_city_bonus(pcity, EFT_CITY_VISION_RADIUS_SQ), 2, 2);
 
-  vision_change_sight(pcity->server.vision, vision_radius_sq);
-  ASSERT_VISION(pcity->server.vision);
+    vision_change_sight(pcity->server.vision, vision_radius_sq);
+    ASSERT_VISION(pcity->server.vision);
+  }
 }
 
 /************************************************************************//**
@@ -3525,8 +3527,11 @@ bool city_map_update_radius_sq(struct city *pcity)
     city_refresh_vision(pcity);
   }
 
-  /* If city is under AI control, update it */
-  adv_city_update(pcity);
+  /* City removal might be ongoing, and advisor data already deleted */
+  if (pcity->server.adv != nullptr) {
+    /* If city is under AI control, update it */
+    adv_city_update(pcity);
+  }
 
   notify_player(city_owner(pcity), city_tile(pcity), E_CITY_RADIUS_SQ,
                 ftc_server, _("The size of the city map of %s is %s."),
-- 
2.43.0

